src/client/Client.cc

```diff
@@ -2083,6 +2083,9 @@ void Client::handle_client_session(MClientSession *m)
 
   case CEPH_SESSION_RECALL_STATE:
     trim_caps(session, m->get_max_caps());
+    session->con->
+    send_message(new MClientSession(CEPH_SESSION_RECALL_STATEMSG_ACK,
```

Comment:
These messages are not necessary which is why I removed them in the previous version. You can just use the flush messages.

Reply:
Okay.

What about [session->crs_ack_rcvd](https://github.com/ceph/ceph/pull/21566/files#diff-23bc98a965757649a7e2d936e1eb7092R104)? Should it stay and where should the [code that checks if all ACKs are received](https://github.com/ceph/ceph/pull/21566/files#diff-ea4df8dd62d8f79137676b9f21d58633R3144) be (not just in textual region but also while execution)?

Reply:
You don't need that but you would want to keep track of which sessions have not dropped caps yet (indicated by which sessions have not sent back `flushmsg_ack` by your timeout). The `Gather` mechanism doesn't have a way to do that. So, you will need to create a subclass of `MDSContextSleeper` which keeps a `std::set<Session *>`. You will also need to virtualize `C_GatherBase::new_sub` and `C_GatherBase::sub_finish`` so you can override these methods in order to track the completions (you will still call `C_GatherBase::new_sub()` in your the overriden method to do the bulk of the work).

Also, consider allowing specifying a timeout via the command argument to `ceph tell cache drop` rather than config option.